### PR TITLE
Hydroponics Mutations Refactor and Rework

### LINKED ISSUE #28624: Feature Request: A way to remove specific chems from a plant's phytochemistry via either a tool or xenobotany machine.
I play botany a lot, and it's always a kick in the nuts to see a great plant that I can't effectively mass use because it has a chem that doesn't help, like the solanine (toxin) in ambrosia vulgaris or the cyticine (synaptizine) in ambrosia deus, both of which make blunts pointless, since smoking a normal blunt makes you somewhat poisoned while a godblunt can kill you (1u cyticine equals about 60 toxdamage and since a godblunt, regardless of potency, has more than 1u, smoking a godblunt is a death sentence.)

Other times, when making materials-fruit (like glowberries for uranium) the presence of nutriment is a problem, since it always outnumbers the material, making any grinding woefully slow since you have to remove the nutriment every time to keep going.
Many times you find the ideal exotic seed, only to find out it not only makes the chems you wanted but also 3 different poisons.

You can ADD chems to a plant via splicing, but there's no way to remove specific chems from it.
Please add a way to remove specific chems from a plant's phytochemistry.
[featurerequest] [qol]

### PR DESCRIPTION
**Pull request by adacovsk**
[system]

The mutations code for hydroponics is awful to read and hopefully this brings about more possibilities. After talking with a couple veteran botanists, I noticed that they used splicing/purging to accomplish most of what they wanted to achieve, avoiding mutations, mainly because of the low odds of getting a mutation that they wanted.

## What this does
- Refactor and redo the whole mutation system
- Rather than mutagen and radium causing all mutations, I broke it up such that:
   - Mutagen affects phytochemistry genes, put teleporting into xenoecology
   - Radium affects morphology genes
   - Uranium affects bioluminescence genes
   - Bicardines affects ecology genes
   - Kelotanes affects ecophysiology genes
   - Hyperzines affects metabolism genes. Metabolism now contains `fluid_consumption` and `nutrient_consumption`
   - Creatine affects development genes
   - Dexalins (Thymol) affects xenoecology. I'll put the very esoteric traits under here, like consume/exude gases and teleporting.

- Removes `custom_plant_metabolism`, `severity`, `yield_mod` and `mutation_mod`: these made barely any impact when used, and were rarely used, and convoluted.
- Rebalanced some toxins in on_plant_life() for some chems
- `parasite` combined with `carnivorous` now known has `voracious` to reduce the number of traits... Level 1 voracious results in the plant eating both weeds and pests. Level 2 `voracious` is HCARN
- Left4Zed actually does something now (2% chance to proc any mutation each tick currently)
- Metabolism no longer exude/consume gases
- Morphology mutations can affect appearance
- Modifies creatine recipe so untable mutagen can count as a reagent in recipe so botanists can produce it.
 - Creatine can be made with Poppies, Ambrosia, glowshrooms, regular and death nettles, but procs more often to compensate
- Somatray affects phytochemistry and development only. This may change. WIP
 
## Why it's good
- This should allow botanists more possibilities with their plants, and hopefully more possibilities for coders in the future as a big limiting factor was usability of the old code
- Mutating phytochemistry has a chance to remove a chemical. Closes #28624. While it is a roundabout way to eliminate chemicals, I don't see an easier way to do it until I modify seed machines (which I'm not doing for this PR).

## Changelog

:cl:
 * tweak: Redid hydroponics mutations, see PR #32730

### CODE REVIEW

**Comment by pedr69** [comment]
>removes mutation_mod

i'm a noob at xenobotany but doesn't this make the left4zed fertilizer literally useless as opposed to now where it's "just" useless

**Comment by adacovsk** [comment]
> > removes mutation_mod
> 
> i'm a noob at xenobotany but doesn't this make the left4zed fertilizer literally useless as opposed to now where it's "just" useless

That's right. I'll make it do something more interesting

**Comment by Eneocho** [comment]
> > removes mutation_mod
> 
> i'm a noob at xenobotany but doesn't this make the left4zed fertilizer literally useless as opposed to now where it's "just" useless

The main effect of L4Z is to make plants WAY more sensitive to mutations (rarer mutations happen more often, stat mutations are more powerful) but not cause mutations by itself. A standard when messing with mutations is to dump 1-2 bottles of L4Z, wait for the chems to go through then somato/mutagen (radium is too weak for mutation gambling) the plant to kingdom come and get all sorts of weird effects.
I kinda dislike that it's getting changed to do one of any mutation type. It should still have the mutation sensitivity effect so you can more easily get the rarer/stronger effects on the respective mutation area.

**Comment by Eneocho** [comment]
>Mutating phytochemistry has a chance to remove a chemical. Closes https://github.com/vgstation-coders/vgstation13/issues/28624

What I was asking for in that issue was to be able to remove specific chemicals, not gamble to remove one of any chem. This is a step forward yeah, but having to gamble 10 times to finally get rid of the 1 (one) chemical you wanted on the plant is a weird choice considering that you're refactoring the thing right now.

**Comment by adacovsk** [comment]
>I kinda dislike that it's getting changed to do one of any mutation type. It should still have the mutation sensitivity effect so you can more easily get the rarer/stronger effects on the respective mutation area.

It's nearly impossible to get specific things easily, like hematophage and kudzu through mutations currently. I get it if you want nothing changed, and this shouldn't affect you much at all.  Just don't use the left4zed for that purpose. It's probably not going to be perfectly balanced yet and will require more testing, but I don't want to change the botanists workflow, just provide them more opportunities. Mutations as they currently are, you have about a 3% chance (if that for some things). With my PR, it will be more like 20%+, with potency even higher, and also I've added even more possible variables to modify (or more can be added in the future).

Severity makes the code very convoluted, hard to read and almost impossible to make any changes. I've eliminated it and kept all the benefits of high severity

>gamble 10 times to finally get rid of the 1 (one) chemical you wanted on the plant is a weird choice considering that you're refactoring the thing right now.

Haven't really touched the seed datums too much yet. I'm working on that. I'll work on figuring out a way for splice and purge to work with removing chemicals so it makes sense.

>get the rarer/stronger effects

I would like to think that no one effect is better than the other and leave that up to the botanist to choose and not ruin their odds. Maybe they want a plant that consumes plasma and generates O2, or generates plasma and consumes CO2, or a carnivorous stinging kudzu. Hopefully the odds of getting any combination they want are the same.

**Comment by Eneocho** [comment]
>Severity makes the code very convoluted, hard to read and almost impossible to make any changes. I've eliminated it and kept all the benefits of high severity

Then all is good, no complaints on my part.

**Comment by Eneocho** [comment]
However, you should change the chem for ecophysiology, since antitox already has a use on plants (reduces toxin damage from using toxic reagents).
And you should restore the Nutrition (not nutrient) genes (instead of merging them with metabolism), they control if the plant is carnivorous, parasitic, hematophage or if it consumes certain gases.

**Comment by Armadingus** [comment]
Eneocho knows his shit, highly recommend listening to him about botany. Also creatine is highly difficult to make en masse, and xenobotany is all about making reagents en masse for those mutations. Maybe consider using ammonia to mutate development? Ammonia is only like EZ nutrient that heals the plant, it'd be nice to see it have more use.

**Comment by Eneocho** [comment]
I'd say Creatine is more tedious than hard to make, since you need to source large amounts of nutriment and doing so in medical is slow if you use the synthmeat method. Here's my personal Creatine recipe for 60u (more than enough for most uses): 20u Carbon, 5u Oxygen, 15u Sugar, 20u Phosphorus, 10 Sulfur, 10 Radium, 10 Chlorine, 10u Nutriment (3x).

**Comment by Eneocho** [comment]
One change I'd like to see is for Robust Harvest to no longer have a [chance to reduce the yield](https://github.com/vgstation-coders/vgstation13/blob/cc75c1d055c54a6ff2b29c9b422c45040b7dcb69/code/modules/reagents/Chemistry-Reagents.dm#L2554), since you sometimes want to top off a splice you made while your potencyplant was getting stronger, only to find that you now only make 3 fruits instead of 5.

**Comment by adacovsk** [comment]
>And you should restore the Nutrition

The issue is that I don't want it to be risk-free for botanists to create really evil plants. If you dump 30u of radium, you're bound to get carnivorous, parasitic, and hematophage, without much effort, and I've combined parasitic and carnivorous, since level 1 was just eating pests. Keep in mind that Splice does not transfer gasses. It at least make it a little more difficult, and with one splice and you can keep everything. Just be careful when dumping cocaine into your plant. Dude, just trust me

> Eneocho knows his shit, highly recommend listening to him about botany.

Yeah, I've been talking to Saul_Myers as well among others that play botany a lot.

>I'd say Creatine is more tedious than hard to make

Development contains `auto_harvest`, `kudzu`, `yield`, `production`, `maturation` all of which are very utilitarian and benefit a syndicate. I want to avoid a 5 minute kudzuloose. Like with Metabolism, I don't want to make it too easy. Also, I've had the somatoray affect phtochemistry and development, rather than just yield and potency, eliminating the tediousness. Just need nettles, death nettles, and glowshrooms (untable mutagen), Ambrosia Deus (cocaine), poppy (opium), and whatever else for nutriment. I've made it so that development doesn't proc on chance.

>using ammonia

I'll see if I can work on dymethylamine and ammonia to make them useful

>I'd like to see is for Robust Harvest to no longer have a [chance to reduce the yield

can do, but it currently does some weird formulas and isn't encapsulated, so I will make it so that it doesn't affect just potency, but phytochemistry instead. The odds to increase potency should be about the same, but will scale slightly differently

**Comment by adacovsk** [comment]
> Eneocho knows his shit, highly recommend listening to him about botany. Also creatine is highly difficult to make en masse, and xenobotany is all about making reagents en masse for those mutations. Maybe consider using ammonia to mutate development? Ammonia is only like EZ nutrient that heals the plant, it'd be nice to see it have more use.



> Eneocho knows his shit, highly recommend listening to him about botany.
Yeah, I've been talking to Saul_Myers as well among others that play botany a lot.

**Comment by Armadingus** [comment]
> 
> I'll see if I can work on dymethylamine and ammonia to make them useful
> 
Diethyl is already extremely useful, it speeds up growth stages, heals, makes the plant hardier and live longer, among other things. It needs no special treatment.

**Comment by Armadingus** [comment]
One more thing, how do you mutate the plant species itself after this rework? Is it still by using mutagen/radium in the early growth stage? Example being, if I wanted to make a tomato into a blue tomato, would you have to still pour mutagen in?

**Comment by adacovsk** [comment]
> One more thing, how do you mutate the plant species itself after this rework? Is it still by using mutagen/radium in the early growth stage? Example being, if I wanted to make a tomato into a blue tomato, would you have to still pour mutagen in?

currently, you can pour any of the reagents in to mutate into a different species, but might consider just making it limited to radium/mutagen

**Comment by Eneocho** [comment]
>The issue is that I don't want it to be risk-free for botanists to create really evil plants. If you dump 30u of radium, you're bound to get carnivorous, parasitic, and hematophage, without much effort, and I've combined parasitic and carnivorous, since level 1 was just eating pests. 

You can just buy some seed crates and get the traits, splice them together in a temporary seed then splice/purge into your target seed, no mutation required. And you don't even need special traits to make evil plants.
Trust me on this one, make the nutrition genes their own thing again, so you can specifically add/remove those traits instead of those AND the development ones. The main gripe I have with the current geneblock system is that you modify the entire chunk instead of being able to pinpoint changes (you splice yield AND production time for example and can't splice one but not the other).
And for the love of god don't make carnivorous and parasitic the same thing, parasitic is feeding on weeds while carnivorous is feeding on pests. They should be separate traits.
The dangerous gene is High Carnivorous (not regular Carnivorous) and that one's only dangerous on vines/kudzu.

>Keep in mind that Splice does not transfer gasses. It at least make it a little more difficult, and with one splice and you can keep everything.

This is a bad change too. Gas type consumption should be a spliceable trait.

>Development contains auto_harvest, kudzu, yield, production, maturation all of which are very utilitarian and benefit a syndicate. I want to avoid a 5 minute kudzuloose. Like with Metabolism, I don't want to make it too easy. Also, I've had the somatoray affect phtochemistry and development, rather than just yield and potency, eliminating the tediousness. Just need nettles, death nettles, and glowshrooms (untable mutagen), Ambrosia Deus (cocaine), poppy (opium), and whatever else for nutriment. I've made it so that development doesn't proc on chance.

5m kudzulooses are possible without mutations, changing the current mutation system won't change that, and balancing the traits and gene system around the idea of kudzuloosing is a very, VERY bad idea. It's like balancing all medicine to be garbage tier shit because you can make mednanos with some effort at roundstart. You're just adding tedium that will not be appreciated.
Also, somato does not permanently affect potency and yield, it affects the current harvest's yield (only the current harvest), but the genes stay the same unless you go for mutation mode (which is the same as using mutagen).
Glowshrooms make uranium, not mutagen (no plant makes mutagen). Uranium, while radioactive, is not mutagenic in game.

>Metabolism mutation can allow for oxygen production rather than just negative gases

But it does so currently. I've made quite a few o2 shitting plants.

**Comment by Eneocho** [comment]
While I do like that you can now specifically mutate a genegroup instead of gambling randomly with mutagen, that mutating morphology changes the appearance and that the somatoray actually affects the genestats instead of tweaking them for the current splice only, I very much dislike the removal of nutrition genes, merging of parasitic and carnivorous traits, balancing of the mutation and trait system around the idea of roundstart kudzuloose (if we're gonna down the very unwise route of balancing shit around the possibility of roundstart LOOSE then we better start removing the singulo, since a syndie can roundstart loose it as well; see how absurd it gets?) and the general tedium for the sake of balance.
Thus far I prefer the current system, despite the issues, than this nu-botany.

**Comment by adacovsk** [comment]
> This is a bad change too. Gas type consumption should be a spliceable trait.

This isn't a change...

>Glowshrooms make uranium, not mutagen (no plant makes mutagen). Uranium, while radioactive, is not a mutagen in game.

They make radium, a reagent in mutagen... Anyway, I've made it so you can create creatine as a botanist and use untable mutagen.

>But it does so currently. I've made quite a few o2 shitting plants.

Not by mutations...

The parasite + carnivorous level 1 combo is a trivial difference. Having a trait separate just for eating plants just reduces your odds for other traits

**Comment by Eneocho** [comment]
>>This is a bad change too. Gas type consumption should be a spliceable trait.

>This isn't a change...

But it is, you can currently splice what gases the plant consumes without issue. I've made a few optimized plasma gas plants for an enterprising atmosian looking for extra oomph for the plasma loop. You're saying that it's no longer a trait that can be spliced, thus a change.

>They make radium, a reagent in mutagen...

I read that as glowberries not glowshrooms, my apologies.

>>But it does so currently. I've made quite a few o2 shitting plants.
>Not by mutations...

Fair enough.

**Comment by Eneocho** [comment]
If you want we can hop in the test serb and I can show you botany firsthand, so you know what your changes affect. I appreciate the effort (apologies if I've gotten defensive, I've lost quite a few fun mechanics because a coder got invested on them), even if I dislike some of the changes.

**Comment by adacovsk** [comment]
> If you want we can hop in the test serb and I can show you botany firsthand, so you know what your changes affect. I appreciate the effort (apologies if I've gotten defensive, I've lost quite a few fun mechanics because a coder got invested on them), even if I dislike some of the changes.

let me know a good time in est

**Comment by Eneocho** [comment]
I mean, if you can join right now I'm available.

**Comment by adacovsk** [comment]
> I mean, if you can join right now I'm available.

ok

**Comment by gurfan** [comment]
>Somatray affects phytochemistry and development only. This may change

Maybe make it so you can use the somatray in hand to select a gene to target

**Comment by Eneocho** [comment]
Consider reworking the noreact trait added in #32716 as well.

**Comment by adacovsk** [comment]
> Consider reworking the noreact trait added in #32716 as well.

yup, conflicts. This is a big WIP

File: code/modules/hydroponics/seeds.dm [hunk 1]
```diff
@@ -1860,6 +1860,6 @@
 	production = 5
 	yield = 5
 	growth_stages = 2
-	fluid_consumption = 0.5
-	nutrient_consumption = 0.15
+	fluid_consumption = 5
+	nutrient_consumption = 1.5
```
Comment (Eneocho) [inline code/modules/hydroponics/seeds.dm:1864]: Why this change?

Comment (adacovsk) [reply code/modules/hydroponics/seeds.dm:1864]: You've probably noticed this before, but hydroponics trays before they were used would show 100/10. With my last hydroponics PR, I scaled nutrient up to 100, so nutrient_consumption would consume 10x more as well

**Comment by Eneocho** [comment]
I'd change the name from "Xenoecology" to "Xenophysiology".
